[feature] : (#88) add twilio template integration

PavanKumarPacharla opened the pull request:
# 🔁 Pull Request: WhatsApp Template Support for Notifications

## 📌 Summary

This PR introduces support for sending **WhatsApp Business Template Messages** via Twilio from ERPNext's Notification system. It includes a new `WhatsApp Message Template` DocType, enhanced configuration options in the Notification UI, and dynamic substitution of template variables during message dispatch.

---

## ✅ Features Added

### 1. DocType: `WhatsApp Message Template`
- Fields:
  - `template_sid`
  - `template_body`
  - Child table: `parameters` (linked to `WhatsApp Template Parameter`)
    - `variable`: Placeholder key
    - `value`: Jinja expression (e.g., `{{ doc.customer_name }}`)

### 2. New DocType: `WhatsApp Template Parameter`
- Used to define parameters for message templates
- Includes:
  - `variable`: the placeholder (e.g., `name`, `date`)
  - `value`: expression rendered using context

### 3. Notification Doctype Enhancements
- New Fields:
  - `use_whatsapp_template` (Checkbox)
  - `whatsapp_template` (Link field)
- UI behavior:
  - If `channel == "WhatsApp"` and `use_whatsapp_template == 1`, hide `message` and show template-related fields.

### 4. WhatsApp Message Dispatch Logic
- If `use_whatsapp_template` is checked:
  - Load the selected `WhatsApp Template`
  - Render parameter values using Jinja and context (`doc`)
  - Replace placeholders in `template_body`
  - Send message via Twilio using `template_sid` and `content_variables`
  - Store rendered `template_body` in Communication and WhatsApp Message logs
- If not using template, send the regular `message`

---

## 🔧 Code Changes

### Backend
- **Modified** `NotificationTwilio.send_whatsapp_msg`
  - Loads template and renders parameters with `frappe.render_template`
  - Updates message sending args accordingly

- **Modified** `WhatsAppMessage.store_whatsapp_message`
  - Stores `template_body` in `message` field if using a template

## 🧪 Testing

- [x] Sending normal WhatsApp messages (non-template)
![image](https://github.com/user-attachments/assets/4cfb7c4e-99f3-43e2-99af-4e7da0cae08f)

- [x] Sending WhatsApp template messages with substituted values
![image](https://github.com/user-attachments/assets/9b573c32-5b90-4314-b381-592bba3fa796)

- [x] Rendered `template_body` stored in Timeline, Communication and WhatsApp Message
![image](https://github.com/user-attachments/assets/367b78d6-c165-4dcc-a763-024d16a58fad)
![image](https://github.com/user-attachments/assets/73abe219-5228-4aca-a8f3-c7f8e76c6657)
![image](https://github.com/user-attachments/assets/c3c99361-28d8-4b4e-984f-3314409a00dd)

- [x] Dynamic UI behavior in Notification Doctype (conditional visibility of fields)
![image](https://github.com/user-attachments/assets/6014bb2f-d910-4f39-b9cc-b853c2cad33b)

Closes: [](https://github.com/ParaLogicTech/frappe/issues/88)

PavanKumarPacharla committed:
[feature] : (#88) add twilio template integration

SaiFi0102 commented on twilio_integration/overrides/notification_hooks.js:
```diff
@@ -1,24 +1,37 @@
+	},
+
+	setup_whatsapp_message_template: function(frm) {
+		const isWhatsapp = frm.doc.channel === 'WhatsApp';
```
use snake_case

PavanKumarPacharla replied:
okay

SaiFi0102 commented on twilio_integration/overrides/notification_hooks.js:
```diff
@@ -1,24 +1,37 @@
+	setup_whatsapp_message_template: function(frm) {
+		const isWhatsapp = frm.doc.channel === 'WhatsApp';
+
+		frm.set_df_property('use_whatsapp_template', 'hidden', !isWhatsapp);
+		frm.set_df_property('whatsapp_message_template', 'hidden', !isWhatsapp || !frm.doc.use_whatsapp_template);
+		frm.set_df_property('message_sb', 'hidden', isWhatsapp && frm.doc.use_whatsapp_template);
```
can be commonified with `toggle_template_fields`

PavanKumarPacharla replied:
i will try to commonify this

SaiFi0102 commented on twilio_integration/overrides/notification_hooks.py:
```diff
@@ -59,20 +60,35 @@ def send_whatsapp_msg(self, doc, context):
+
+			message = template.get_rendered_body(context)
+
+			args["message"] = None
+			args["template_sid"] = template.template_sid
+			args["template_body"] = message
```
why keep message and template body as separate fields

PavanKumarPacharla replied:
To Store the body in whatsapp message and id to send to the twilio client

SaiFi0102 commented on twilio_integration/overrides/notification_hooks.py:
```diff
@@ -59,20 +60,35 @@ def send_whatsapp_msg(self, doc, context):
+		}
+
+		if self.use_whatsapp_template and self.whatsapp_template:
+			template = frappe.get_doc("WhatsApp Message Template", self.whatsapp_message_template)
```
Use `frappe.get_cached_doc` instead

PavanKumarPacharla replied:
okay i will modify that

SaiFi0102 commented on twilio_integration/twilio_integration/doctype/whatsapp_message_template/whatsapp_message_template.json:
```diff
@@ -7,34 +7,47 @@
  "modified_by": "Administrator",
  "module": "Twilio Integration",
  "name": "WhatsApp Message Template",
+ "naming_rule": "By fieldname",
```
where is `field:template_name`?

PavanKumarPacharla replied:
Earlier while we are discussing we have seen the doctype i created "WhatsApp Template", in that we modified this naming field, but later i realized as we already having "WhatsApp Message Template" doctype and use this doctype and naming rule already exists as template_name, hence it is not showing in git changes

PavanKumarPacharla committed:
[feature] : (#88) PR suggested changes

SaiFi0102 committed:
fix(PR Review): Missing custom fields

SaiFi0102 committed:
fix: PR Review changes
- Missing fields in WhatsApp Message + read only fields
- Explicit method call for readability
